gflownet/envs/crystals/ccrystal.py

```diff
@@ -0,0 +1,917 @@
+        """
+        return self.statebatch2oracle(states)
+
+    def statetorch2proxy(
+        self, states: TensorType["batch", "state_dim"]
+    ) -> TensorType["batch", "state_oracle_dim"]:
+        """
+        Returns statetorch2oracle(states).
+        """
+        return self.statetorch2oracle(states)
+
+    def set_state(self, state: List, done: Optional[bool] = False):
+        super().set_state(state, done)
+
+        stage_idx = self._get_stage(state).value
+
+        # Determine which subenvs are done based on stage and done
+        done_subenvs = [True] * stage_idx + [False] * (len(self.subenvs) - stage_idx)
+        done_subenvs[-1] = done
+        # Set state and done of each sub-environment
+        for (stage, subenv), subenv_done in zip(self.subenvs.items(), done_subenvs):
+            subenv.set_state(self._get_state_of_subenv(state, stage), subenv_done)
+
+        """
+        We synchronize LatticeParameter's lattice system with the one of SpaceGroup
+        (if it was set) or reset it to the default triclinic otherwise. Why this is 
+        needed: for backward sampling, where we start from an arbitrary terminal state,
+        and need to synchronize the LatticeParameter's lattice system to what that
+        state indicates,
+        """
+        if self.subenvs[Stage.SPACE_GROUP].done:
+            lattice_system = self.subenvs[Stage.SPACE_GROUP].lattice_system
+            if lattice_system != "None":
+                self.subenvs[Stage.LATTICE_PARAMETERS].set_lattice_system(
+                    lattice_system
+                )
+            else:
+                self.subenvs[Stage.LATTICE_PARAMETERS].set_lattice_system(TRICLINIC)
+        # Set stoichiometry constraints in space group sub-environment
+        if self.do_stoichiometry_sg_check and self.subenvs[Stage.COMPOSITION].done:
+            self.subenvs[Stage.SPACE_GROUP].set_n_atoms_compatibility_dict(
+                self.subenvs[Stage.COMPOSITION].state
+            )
+
+    def state2readable(self, state: Optional[List[int]] = None) -> str:
+        if state is None:
+            state = self.state
+
+        readables = [
+            subenv.state2readable(self._get_state_of_subenv(state, stage))
+            for stage, subenv in self.subenvs.items()
+        ]
+        return (
+            f"{self._get_stage(state)}; "
+            f"Composition = {readables[0]}; "
+            f"SpaceGroup = {readables[1]}; "
+            f"LatticeParameters = {readables[2]}"
+        )
+
+    def process_data_set(self, data: List[List]) -> List[List]:
+        is_valid_list = []
+        for x in data:
+            is_valid_list.append(
+                all(
+                    [
+                        subenv.is_valid(self._get_state_of_subenv(x, stage))
+                        for stage, subenv in self.subenvs.items()
+                    ]
+                )
+            )
+        return [x for x, is_valid in zip(data, is_valid_list) if is_valid]
```

Comment:
Why not just have a list of results, and if given data point is valid in the for loop, append it to that list?

Reply:
because that uses up to double the amount of memory no? (currently the process gets killed pretty often so any savings counts)

Reply:
I don't see why (but maybe I am missing something)? Both
```
result = []
for x in data:
    if all(
        [
            subenv.is_valid(self._get_state_of_subenv(x, stage))
            for stage, subenv in self.subenvs.items()
        ]
    ):
        result.append(x)
return result
```
and your version keep both the original `data`, and a list of `# valid` data points; your just creates it at the end of the function?

Reply:
Genuine question: does this
```
return [x for x, is_valid in zip(data, is_valid_list) if is_valid]
```
create copies of the x's in memory? If so I agree and then the more memory-efficient way would be with `del`, no?

Reply:
If `x`'s are lists I believe neither version copies them, it just stores a reference to the original object. Example:
```
>>> data = [[1, 2, 3], [4, 5, 6]]
>>> valid = []
>>> valid.append(data[0])
>>> valid
[[1, 2, 3]]
>>> valid[0][0] = 12
>>> valid
[[12, 2, 3]]
>>> data
[[12, 2, 3], [4, 5, 6]]
```

Reply:
I see. Then both are equivalent and "your way¨ is more intuitive indeed.

Reply:
By the way, so is this actually causing memory issues? How big this dataset is? Because from what I see this is done only once (not per copy of an environment)?